### DISCUSSION TITLE
Possible reason for why fluidsynth is saying 'Not a DLS file (RIFF id = 'sfbk♦')'?

### DISCUSSION THREAD

**Discussion by ghost** [FluidSynth/fluidsynth#1084] [Q&A] [answered]
The example fluidsynth https://www.fluidsynth.org/api/example_8c-example.html (compiled with v2.2.5) perfectly runs on my Windows system and my other unix system without producing an error (with sf2 https://github.com/dats-lang/dats-src/blob/master/share/soundfonts/TimGM6mb.sf2).

On a software I wrote which uses fluidsynth, it gives the following error on Windows:

```
fluidsynth: error: Preset header indices not monotonic
fluidsynth: error: Couldn't parse presets from soundfont file

** (dats.exe:12832): CRITICAL **: file D:\a\1\s\libinstpatch\IpatchContainer.c: line 638: assertion `IPATCH_IS_CONTAINER(container)' failed
fluidsynth: error: ipatch_dls_reader_new() failed with error: 'Not a DLS file (RIFF id = 'sfbk♦')'
fluidsynth: error: Failed to load SoundFont "default.sf2"
```

even though the software I wrote works on my other unix system. It happens when `fluid_synth_sfload` is executed.

Here's the first 160 bytes of the sf2:
```
00000000  52 49 46 46 74 17 5b 00  73 66 62 6b 4c 49 53 54  |RIFFt.[.sfbkLIST|
00000010  50 00 00 00 49 4e 46 4f  69 66 69 6c 04 00 00 00  |P...INFOifil....|
00000020  02 00 01 00 49 4e 41 4d  0e 00 00 00 54 69 6d 47  |....INAM....TimG|
00000030  4d 36 6d 62 31 2e 73 66  32 00 69 73 6e 67 08 00  |M6mb1.sf2.isng..|
00000040  00 00 45 4d 55 38 30 30  30 00 49 53 46 54 12 00  |..EMU8000.ISFT..|
00000050  00 00 41 77 61 76 65 20  53 74 75 64 69 6f 20 76  |..Awave Studio v|
00000060  38 2e 35 00 4c 49 53 54  fc f4 57 00 73 64 74 61  |8.5.LIST..W.sdta|
00000070  73 6d 70 6c f0 f4 57 00  00 00 00 00 00 00 00 00  |smpl..W.........|
00000080  00 00 00 00 00 00 ff ff  fb ff f4 ff f0 ff f0 ff  |................|
00000090  f3 ff f2 ff f0 ff ee ff  f0 ff f2 ff f3 ff fd ff  |................|
```

**Comment by derselbst** [comment]
I cannot reproduce this by just using fluidsynth's commandline tool on Windows. I need some more information:

- What Windows are you exactly using? (Version and architecture)
- What "other unix system" are you referring to?

> On a software I wrote which uses fluidsynth

- How does this software look like? Pls. post some code which shows the usage of fluidsynth by that software
- Are you also using fluidsynth 2.2.5 there or a different version?
- Can you reproduce that error with other soundfonts as well?
- Does it fail reproducibly or randomly?

**Reply by ghost** [reply]
> * What Windows are you exactly using? (Version and architecture)

I'm using Windows 10  version 1903. x64 based PC

> * What "other unix system" are you referring to?

On my WSL and on the Termux android application,.

> > On a software I wrote which uses fluidsynth
> * How does this software look like? Pls. post some code which shows the usage of fluidsynth by that software

It's pretty simple, here's an excerpt.
```
static int synth(const symrec_t *const staff, track_t *const pcm_ctx) {
  char *sf2_name;
    sf2_name = "C:\\soundfonts\\default.sf2";
    printf("using %s\n", sf2_name);

  fluid_settings_t *settings;
  fluid_synth_t *lsynth = NULL;
  int err = 0;
  /* Create the settings object. This example uses the default
   * values for the settings. */
  settings = new_fluid_settings();
  if (settings == NULL) {
    fprintf(stderr,"Failed to create the settings\n");
    err = 2;
    if (lsynth) {
      delete_fluid_synth(lsynth);
    }
    if (settings) {
      delete_fluid_settings(settings);
    }
    return 1;
  }
  /* Create the synthesizer */
  lsynth = new_fluid_synth(settings);
  if (lsynth == NULL) {
    fprintf(stderr,"Failed to create the synthesizer\n");
    err = 3;
    if (lsynth) {
      delete_fluid_synth(lsynth);
    }
    if (settings) {
      delete_fluid_settings(settings);
    }
    return 1;
  }

  /* ERROR APPEARS when fluid_synth_sfload is executed */
  /* Load the soundfont */
  if (fluid_synth_sfload(lsynth, sf2_name, 1) == -1) {
    fprintf(stderr,"Failed to load the SoundFont\n");
    err = 4;
    if (lsynth) {
      delete_fluid_synth(lsynth);
    }
    if (settings) {
      delete_fluid_settings(settings);
    }
    return 1;
  }
  /* We're done here */
  if (lsynth) {
    delete_fluid_synth(lsynth);
  }
  if (settings) {
    delete_fluid_settings(settings);
  }
  reset_options_to_default();

  return 1;
}
```
Since I see no problem in the source, I suspect the problem comes from the thousand line of codes I wrote which is incompatible
with Windows. Looking at them line by line isn't so smart at all so I instead asked here why would `fluid_synth_sfload` give this error.

> * Are you also using fluidsynth 2.2.5 there or a different version?

I've tested this on fluidsynth 2.2.5 (windows) and fluidsynth v2.2.2 (windows and unix).

> * Can you reproduce that error with other soundfonts as well?

Tring to use my "MuseScore_General.sf3" it gives:
```
fluidsynth: error: Invalid chunk id in level 0 parse

** (dats.exe:12724): CRITICAL **: file D:\a\1\s\libinstpatch\IpatchContainer.c: line 638: assertion `IPATCH_IS_CONTAINER(container)' failed
fluidsynth: error: ipatch_dls_reader_new() failed with error: 'Not a DLS file (RIFF id = 'sfbk♦')'
fluidsynth: error: Failed to load SoundFont "MuseScore_General.sf3"
```

and trying to use Nice-Steinway-Lite-v3.0.sf2 https://drive.google.com/file/d/1e5nPbx7_yPY6mNr3H1pInUgTKmGM6cHo/view?usp=sharing (link from https://sites.google.com/site/soundfonts4u/) gives:

```
fluidsynth: error: Invalid chunk id in level 0 parse

** (dats.exe:756): CRITICAL **: file D:\a\1\s\libinstpatch\IpatchContainer.c: line 638: assertion `IPATCH_IS_CONTAINER(container)' failed
fluidsynth: error: ipatch_dls_reader_new() failed with error: 'Not a DLS file (RIFF id = 'sfbk♦')'
fluidsynth: error: Failed to load SoundFont "Nice-Steinway-Lite-v3.0.sf2"
```
> * Does it fail reproducibly or randomly?

It's consistent with it's failure no matter how many times I run the software .

**Reply by derselbst** [reply]
Thanks. The source code looks ok to me. And I'm on the same Windows and cannot reproduce it either. Because of that, it's hard to tell what going wrong. I can only explain what's happening on a high level:

Fluidsynth tries to load the given soundfont(s) and finds that they seem to be invalid or corrupted. That's where the error messages like

> Invalid chunk id in level 0 parse

and 

> Preset header indices not monotonic

are coming from. Those errors are raised by the default soundfont loader, and they are fatal. Because of that, loading the font is given to the next soundfont loader "libinstpatch", which believes it is a DLS file, which obviously is not, and then outputs this non-sense message "Not a DLS file".

You need to figure out why the fonts are corrupted once fluidsynth's default loader tries to load them. Maybe start be setting a breakpoint in `read_listchunk()` and see what's the chunk ID fluidsynth has read and expected.

**Reply by ghost** [reply]
Are there perhaps a fluidsynth (for v2.2.5) compiled with debugging symbols with -ggdb3?  I can't seem to do `b read_listchunk` with gdb. Doing `strings libfluidsynth-3.dll | grep read_listchunk` doesn't result any.

**Reply by derselbst** [reply]
> Are there perhaps a fluidsynth (for v2.2.5) compiled with debugging symbols with -ggdb3?

I'm afraid not. You'll have to compile it yourself. It's pretty straightforward on Windows when using vcpkg: `vcpkg install glib:x64-windows`

**Comment by ghost** [comment, marked as answer]
On a completely unrelated reasons and complete unrelated `FILE *`, I had to close other unneeded FILE stream for fluidsynth to work,

https://github.com/dats-lang/dats-src/blob/69949662fb6f28a07e8af85b2023bf5acb099176/dats/semantic.c#L190-#L191

Finally working https://files.catbox.moe/7rps4e.mp3
